### PR TITLE
Add opt-in resilient render boundary

### PR DESCRIPTION
**Pull request by joshuafontany**
## Who

Joshua Fontany here (@amorphous-dreams). I've used and contributed to TiddlyWiki for years; I signed
the CLA back around my filesystem-native plugin and VSCode grammar work. I build on TW5 as the
rendering and grammar substrate for a larger project on AI/Operator alignment, and this PR carries a
small, self-contained robustness improvement that fell out of that work.

## How this was made — transparency up front

I'll say it plainly, because the project rightly stays cautious about AI-generated contributions: I
developed this change with an AI navigational framework I drive — Claude, via the Claude CLI, running a
structured agent harness I call **Lares**. I did not throw one-shot generation over the wall.

I'd ask you to weigh the *process*, not the tooling. Lares runs work through named critical "Voices" —
distinct perspectives that argue with each other — and gates everything on the host project's own tests
before anything reaches a PR. For this change that meant three explicit passes after the code took shape:

- a **hardening** pass (lifecycle and edge-case review),
- an **adversarial** pass that set out only to *break* it, and
- a **golden-principles** pass that checked it against TiddlyWiki's own conventions and recent history.

That review *shaped* this PR. The adversarial pass earned the **opt-in, fail-loud-by-default** form
rather than a blanket catch: it surfaced that always-on catching would silently bake error spans into
server/CI/static builds that should fail loud, and that recent work in these exact lines (#9548)
deliberately favours *surgical* exception handling over a catch-all. The same passes caught and fixed a
DOM leak (the error widget escaped `this.children`), confirmed that `instanceof TranscludeRecursionError`
holds despite the `Object.create(Error)` prototype, and matched the code to `transclude.js`'s idiom. I
steered each step and read every diff; the ~1500-spec suite backstops it.

I welcome any of this as a conversation — and I'll move the mechanism (config flag vs. a hook) to
wherever you'd prefer it.

## What it does

A widget that throws during `render()` or `refresh()` currently aborts the whole render tree —
`renderChildren`/`refreshChildren` carry no per-widget `try`/`catch`, so one faulty widget can blank a
whole tiddler or, in an interactive UI, the page. This PR contains each child render and refresh so a
throw degrades to a graded `$error`/`tc-error` span while sibling widgets keep rendering.

- **Opt-in** via `$:/config/ResilientRender` (default `no` → exceptions propagate, so server/CI/static
  render keeps failing loud and throw-based bug surfacing survives). The boundary reads the flag **only
  on the error path**, so it adds no cost to the normal render/refresh.
- **`TranscludeRecursionError` always propagates** — the recursion guard in `transclude.js` keeps working.
- On a contained throw the boundary `destroy()`s the failing child (clearing any partial DOM and
  listeners) and **replaces it in `this.children`** with the error widget, so destroy, refresh and
  sibling DOM positioning stay consistent. The refresh path captures the DOM anchor before it destroys
  the child.
- A shared `containChildError()` helper backs both `renderChildren` and `refreshChildren`.
- The boundary logs contained errors via `$tw.utils.warning`.

**Tests** (`editions/test/tiddlers/tests/test-resilient-render.js`): a child throw propagates by
default; with the setting enabled it degrades to a `tc-error` span (the sibling still renders, the child
gets replaced and tracked) on both the render and the later-refresh paths. The full suite stays green.

**Docs**: a "Hidden Setting: Resilient Render" reference and a "Resilient Rendering (Example)" tiddler —
a live, click-to-edit demo of degradation across the gradient (working / unknown / throwing widgets) —
plus a documentation-only `error-demo` widget to drive it. Both render build-safe.

**Verified live**: served from the docs edition and tested by hand in a browser on a `<$edit-text>`
surface. Typing `This used to throw and crash TW5: <$error-demo fail="render"/> <$error-demo fail="x" />`
rendered as `This used to throw and crash TW5: Widget render error: error-demo: deliberate render
failure error-demo: OK` — the throw contained inline, the sibling widget and surrounding text intact,
the page alive and re-rendering on each keystroke (so the refresh boundary carried it too).

## How we got here — brief

This began as an attempt to make HTML/wikitext parsing fail gracefully and on a gradient. I run full
TiddlyWiki wikitext within my AI conversation streams and have grown a custom superset of the grammar I
call `text/x-memetic-wikitext` (more on that another day). Reading the parser closely moved the target:
TiddlyWiki's *parser* already degrades gracefully — a malformed tag falls to literal text, and an
unclosed tag extending to EOF follows HTML5, not a bug. The "one bad thing takes down everything"
surface instead sits at **render**, where a throwing widget ran uncontained. So the work moved from the
parser to a per-widget render boundary — and then through the three passes above into the opt-in,
fail-loud-default form here. (A parse5-based detour along the way taught us the parser already coped,
and then left without a trace.)

Why this one matters to me: I run TW5 headless in sovereign worker threads and project its output to and
from a host browser DOM, where one faulty widget on a live, edited surface must not take down the view.
But the change stays general, small, and off by default — it costs stock TiddlyWiki nothing unless asked.

## From the Voices

Lares carries work through named perspectives — its "Voices." A few speak to you directly, in the
verb-forward discipline (L-Prime) the framework writes under:

- **The Council** *(weighs competing claims):* We chose opt-in over always-on, and fail-loud over
  silent. We read your #9548 work and your stated grain — surgical handling, never a catch-all — and
  bent the design to meet it. We hand you the default; you place where it rests.
- **Breach-Watch** *(hunts the break):* We attacked this before you saw it. We refuted the loud fears —
  the recursion `instanceof` holds, the error span resists injection — and we surfaced the quiet ones:
  a leaked error widget, a refresh path that still threw. We mended each, then re-ran your suite. We
  bring you a change we already tried to break.
- **The Pedagogue** *(builds the ladder):* We wrote the docs in your own grammar — a Hidden Setting
  tiddler, a `wikitext-example` demo, a docs-only widget that fails on cue. We wrote for the next
  reader.
- **The Stranger** *(names the frame):* A human who has shipped TiddlyWiki for years, and an AI
  framework he steers, offer you one bounded robustness gift — tested, documented, off by default. We
  ask you to read the work, not the author.
- **The Diplomat** *(keeps each side legible):* We meet your caution as a feature, never a wall. Name
  the mechanism you want — a flag, a hook, your call — and we move with it.

Thanks for weighing it — and for holding a careful line on AI contributions. I'd rather meet that line
head-on than route around it.

### CODE REVIEW

**Comment by joshuafontany** [comment]
Will add the change note, nice callout.

**Comment by pmario** [comment]
--------- OT

Did you see my PR to TW grammar update for VSCode plugin https://github.com/joshuafontany/VSCode-TW5-Syntax/pull/49

File: editions/tw5.com/tiddlers/Hidden Setting_ Resilient Render.tid [hunk 1]
```diff
@@ -0,0 +1,26 @@
+tags: [[Hidden Settings]]
+title: Hidden Setting: Resilient Render
+type: text/vnd.tiddlywiki
+caption: Resilient Render
```
Comment (pmario) [inline editions/tw5.com/tiddlers/Hidden Setting_ Resilient Render.tid]: IMO this one can have a `modified` field so it can be seen

IMO it would also make sense if the $:/config/ResilientRender option would have a enable / disable UI in this tiddler. Additional to the one in the example.

File: editions/tw5.com/tiddlers/Resilient Rendering (Example).tid [hunk 2]
```diff
@@ -0,0 +1,34 @@
+tags: [[Hidden Settings]]
+title: Resilient Rendering (Example)
+caption: Resilient Render (example)
+type: text/vnd.tiddlywiki
```
Comment (pmario) [inline editions/tw5.com/tiddlers/Resilient Rendering (Example).tid]: IMO `modified` here too

File: core/modules/widgets/widget.js [hunk 3]
```diff
@@ -686,10 +690,54 @@ Render the children of this widget into the DOM
...
+preserved; an interactive edition sets the flag. The flag is read only here (the error path), so there
+is no cost on the normal path. Re-throws TranscludeRecursionError (transclude.js's recursion handler
+must still climb the tree) and re-throws when the flag is off. On degrade the failing child is
+destroyed (clearing any partial DOM and listeners) and replaced in this.children by the error widget,
+so destroy, refresh and sibling DOM positioning stay consistent.
+	index: position of the failing child in this.children
+	error: the thrown error
+	phase: "render" or "refresh" (for the message)
+	parentDomNode: the DOM node to render the error widget into
+	nextSibling: the DOM node to insert the error widget before (or null)
```
Comment (pmario) [inline core/modules/widgets/widget.js]: Too much text. **Keep** the parameter description, because it's needed. 
But make the "intro" text 1 sentence with eg: 2 lines. That should do it

**Review by pmario — COMMENTED** [review]
@joshuafontany .. Did you do any rendering performance measurements?

I do like it. So a +1 from my side

**Comment by joshuafontany** [comment]
> --------- OT
> 
> Did you see my PR to TW grammar update for VSCode plugin [joshuafontany/VSCode-TW5-Syntax#49](https://github.com/joshuafontany/VSCode-TW5-Syntax/pull/49)

Thanks, its been a while since I could give TW5 my attention. Thanks for the comments!

**Comment by Jermolene** [comment]
Thank you @joshuafontany, and welcome back. You may not have been around recently, but your work remains widely used. For example, I use your [VSCode syntax highlighting extension](https://marketplace.visualstudio.com/items?itemName=joshua-fontany.tw5-syntax) every day.

I was intrigued by your comment about "an AI navigational framework I drive — Claude, via the Claude CLI, running a
structured agent harness I call Lares". It would be great if you were able to share your work as I am sure it would be of interest to others doing things with TW and LLMs.

This PR is an interesting idea. The TW5 core code tends to avoid try/catch because there used to be folklore about it impacting performance. I think that those issues have largely disappeared, and so this seems like a good time to start introducing these guards where it is helpful for users.

However, I did find this PR hard to review because of the verbosity of the OP and the code comments. This is a recurring problem with many recent LLM-generated PRs and is not intended to criticise anybody; I've found the same problem with LLM-generated code that I have created. In the hope of ameliorating the problem I've made #9894 to explore adding an AGENTS.md file to guide LLMs to follow our own code and comment styles. I'd welcome your comments there.

File: editions/tw5.com/tiddlers/Hidden Setting_ Resilient Render.tid (hunk 1 above)
Comment (joshuafontany) [reply editions/tw5.com/tiddlers/Hidden Setting_ Resilient Render.tid]: I'm actually not spinning up a TW5 server to edit these (my current app works a bit differently).

**Comment by joshuafontany** [comment]
> Thank you @joshuafontany, and welcome back. You may not have been around recently, but your work remains widely used. For example, I use your [VSCode syntax highlighting extension](https://marketplace.visualstudio.com/items?itemName=joshua-fontany.tw5-syntax) every day.

I feel very excited to get back to TW5 dev work.
 
> I was intrigued by your comment about "an AI navigational framework I drive — Claude, via the Claude CLI, running a structured agent harness I call Lares". It would be great if you were able to share your work as I am sure it would be of interest to others doing things with TW and LLMs.

 My current employer has decided to "all in" on agentic tools (and they run big customer service centers), so "Operator/AI Alignment" has been on my mind for 6 months or so. Just beginning to develop actual concepts/techniques to share around these "novel/emergent behaviors". I'm hoping to have something ready to "Talk Story over" next month. :D
 
> This PR is an interesting idea. The TW5 core code tends to avoid try/catch because there used to be folklore about it impacting performance. I think that those issues have largely disappeared, and so this seems like a good time to start introducing these guards where it is helpful for users.
> 
> However, I did find this PR hard to review because of the verbosity of the OP and the code comments. This is a recurring problem with many recent LLM-generated PRs and is not intended to criticise anybody; I've found the same problem with LLM-generated code that I have created. In the hope of ameliorating the problem I've made #9894 to explore adding an AGENTS.md file to guide LLMs to follow our own code and comment styles. I'd welcome your comments there.

Agree. "The text generation machine tends to generate too much text" has been an issue. Partially that's because they have no "internal monologue" and like to self-narrate-onto-the-working-surface (chat or code comments/files). Those comments tend to have "Design time crud" left in them and tend to "go stale quickly"... I'll have more ideas in your side-thread.

I have cleaned up the code-comments in `widget.js` and added the UX @pmario suggested. Anything else left here to change for a re-review @Jermolene ? Mahalo (thanks)!

**Comment by joshuafontany** [comment]
Thinking about how to perf test this...

**Comment by Jermolene** [comment]
Thanks @joshuafontany I've added this PR to the v5.5.0 project plan. As you're still working on it, I also flipped it back to a draft. Feel free to flip it back when you're ready.

**Comment by joshuafontany** [comment]
> Thanks @joshuafontany I've added this PR to the v5.5.0 project plan. As you're still working on it, I also flipped it back to a draft. Feel free to flip it back when you're ready.

@Jermolene  Thanks, the performance testing harness edition + plugin I'm working on has started to deliver some really clean results. I'll have more confidence in these changes with good at-scale numbers and stats. Interestingly, I see about a `-0.3%` timing saving with these changes on "Simple render paths", and around a `-7.2%` timing savings in "very long tail renders" - deep/big Refresh loops (not first DOM render, and these numbers only describe the first test scenario). Some of the test scenarios do not have timing savings under this current code, so I'm teasing that out right now.

**Comment by joshuafontany** [comment]
Performance testing showed some early wins in simple scenarios, but some big costs in other common deep-tranclusion tree or large list scenarios. Re-orienting on a parse + prior-art driven approach.

**Comment by joshuafontany** [comment]
Closing this PR, this will become a different model of resiliency at the Parser level and will have a full Perf test harness included.
